File: src/303. Range Sum Query - Immutable.js
export class NumArray {
  /**
   * @param {number[]} nums
   */
  constructor(nums) {
    this.prefixSums = Array(nums.length + 1)
    this.prefixSums[0] = 0
    for (let i = 1; i < this.prefixSums.length; i++) {
      this.prefixSums[i] = this.prefixSums[i - 1] + nums[i - 1]
    }
  }
  /**
   * @param {number} left
   * @param {number} right
   * @return {number}
   */
  sumRange(left, right) {
    return this.prefixSums[right + 1] - this.prefixSums[left]
  }
}

/**
 * Your NumArray object will be instantiated and called as such:
 * var obj = new NumArray(nums)
 * var param_1 = obj.sumRange(left,right)
 */

File: src/303. Range Sum Query - Immutable.cpp
class NumArray {
private:
    vector<int> prefixSums;
public:
    NumArray(vector<int>& nums) {
        int n = nums.size();
        prefixSums.resize(n + 1);
        for (int i = 1; i < prefixSums.size(); ++i) {
            prefixSums[i] = prefixSums[i - 1] + nums[i - 1];
        }
    }
    int sumRange(int left, int right) {
        return prefixSums[right + 1] - prefixSums[left];
    }
};

/**
 * Your NumArray object will be instantiated and called as such:
 * NumArray* obj = new NumArray(nums);
 * int param_1 = obj->sumRange(left,right);
 */